Sources/AppcuesKit/Presentation/Extensions/UIScrollView+ScrollObserver.swift

```diff
@@ -0,0 +1,237 @@
+        // in and override the scrollViewDidScroll callback and monitor for UI scroll changes that might
+        // impact Appcues experience rendering
+        let originalScrollViewDelegateSelector = #selector(getter: self.delegate)
+        let swizzledScrollViewDelegateSelector = #selector(appcues__getScrollViewDelegate)
+
+        guard let originalScrollViewMethod = class_getInstanceMethod(self, originalScrollViewDelegateSelector),
+              let swizzledScrollViewMethod = class_getInstanceMethod(self, swizzledScrollViewDelegateSelector) else {
+            return
+        }
+
+        method_exchangeImplementations(originalScrollViewMethod, swizzledScrollViewMethod)
+    }
+
+    // this is our custom getter logic for the UIScrollView.delegate
+    @objc
+    private func appcues__getScrollViewDelegate() -> UIScrollViewDelegate? {
+        let delegate: UIScrollViewDelegate
+
+        var shouldSetDelegate = false
+
+        // this call looks recursive, but it is not, it is calling the swapped implementation
+        // to get the actual delegate value that has been assigned, if any - can be nil
+        if let existingDelegate = appcues__getScrollViewDelegate() {
+            delegate = existingDelegate
+        } else {
+            // if it is nil, then we assign our own delegate implementation so there is
+            // something hooked in to listen to scroll
+            delegate = AppcuesScrollViewDelegate.shared
+            shouldSetDelegate = true
+        }
+
+        swizzle(
+            delegate,
+            targetSelector: NSSelectorFromString("scrollViewDidScroll:"),
+            placeholderSelector: #selector(appcues__placeholderScrollViewDidScroll),
+            swizzleSelector: #selector(appcues__scrollViewDidScroll)
+        )
+
+        swizzle(
+            delegate,
+            targetSelector: NSSelectorFromString("scrollViewDidEndDecelerating:"),
+            placeholderSelector: #selector(appcues__placeholderScrollViewDidEndDecelerating),
+            swizzleSelector: #selector(appcues__scrollViewDidEndDecelerating)
+        )
+
+        swizzle(
+            delegate,
+            targetSelector: NSSelectorFromString("scrollViewDidEndDragging:willDecelerate:"),
+            placeholderSelector: #selector(appcues__placeholderScrollViewDidEndDragging),
+            swizzleSelector: #selector(appcues__scrollViewDidEndDragging)
+        )
+
+        // If we need to set a non-nil implementation where there previously was not one,
+        // swap the swizzled getter back first, then assign, then restore the swizzled getter.
+        // This is done to avoid infinite recursion in some cases observed, where a UICollectionView,
+        // for example, may call the getter during the execution of the setter.
+        if shouldSetDelegate {
+            UIScrollView.swizzleScrollViewGetDelegate()
+            self.delegate = delegate
+            UIScrollView.swizzleScrollViewGetDelegate()
+        }
```

Comment:
Calling out this one particularly tricky spot - I found that without this, UICollectionVIew usage would crash the app with the infinite loop.

At first, I thought simply returning `delegate` (L138 below) would be sufficient to "fake" the assignment of a delegate, in the case where the host app had no delegate assigned to swizzle into - but it was not. Thus, the assignment on L134 of our default `AppcuesScrollViewDelegate.shared` (from L103 above) seems to be required.